File: src/ui/managers/message_manager.py
from PySide6.QtCore import Qt

from shared_toolkit.ui.message_dialog import MessageKind, open_non_modal_message


class MessageManager:
    def __init__(self, host):
        self.host = host

    def show_non_modal_message(self, kind: MessageKind, title: str, text: str):
        dialog = open_non_modal_message(
            self.host.parent_widget,
            kind=kind,
            title=title,
            text=text,
        )

        self.host._active_message_boxes.append(dialog)
        dialog.finished.connect(
            lambda: self.host._active_message_boxes.remove(dialog)
            if dialog in self.host._active_message_boxes
            else None
        )

File: cpp/app/ui/managers/message_manager.cpp
#include "ui/managers/message_manager.h"

namespace imgsli::app::ui::managers {

MessageManager::MessageManager(QWidget* parent)
    : QObject(parent), parentWidget_(parent) {}

void MessageManager::showNonModal(QMessageBox::Icon icon, const QString& title,
                                  const QString& text) {
  auto* box = new QMessageBox(parentWidget_);
  box->setWindowFlags(box->windowFlags() | Qt::Window);
  box->setAttribute(Qt::WA_DeleteOnClose);
  box->setModal(false);
  box->setIcon(icon);
  box->setWindowTitle(title);
  box->setText(text);

  active_.append(box);
  connect(box, &QMessageBox::finished, this, [this, box]() {
    active_.removeAll(box);
  });

  box->show();
  box->raise();
  box->activateWindow();
}

}  // namespace imgsli::app::ui::managers
